### PR TITLE
feat: merge `bundleMDX` and `bundleMDXFile`

### PR DESCRIPTION
**Pull request by Arcath**
#121 asked about using the frontmatter in configuration which sounds like a reasonable request. `bundleMDX` couldn't provide this feature as to read the frontmatter it needs to know if its reading a file or source which it can't do until after the options object has been built, so it can't then provide the frontmatter to the options functions.

This PR changes to using a `source` or `file` property on the options argument and doing away with the first argument.

```ts
const mdxSource = `
---
title: foo
---

Some content
`

// old
bundleMDX(mdxSource, {
  // some options
})

// new
bundleMDX({
  source: mdxSource,
  // some options
})

const filePath = path.join(process.cwd(), 'sample.mdx')

//old
bundleMDXFile(fiilePath, {
  // some options
})

// new
bundleMDX({
  file: filePath,
  // some options
})
```

I've also added the features for #121 where `xdmOptions` and `esbuildOptions` are given a second argument of the frontmatter.

Still to do:

 - [x] Document the frontmatter feature
 - [x] Type the frontmatter better. (Maybe `bundleMDX<Frontmatter>()>`)
 - [x] Similar setup for the output.

Ideally I'd like to take the _common_ tasks of chaning input/output and just make them options.
**Checklist**:

- [x] Documentation
- [x] Tests
- [ ] Ready to be merged

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -45,7 +45,7 @@
     "@fal-works/esbuild-plugin-global-externals": "^2.1.2",
     "gray-matter": "^4.0.3",
     "remark-frontmatter": "^4.0.1",
-    "remark-mdx-frontmatter": "^1.0.1",
+    "remark-mdx-frontmatter": "1.0.1",
```
Comment (Arcath) [inline package.json:48]: There appears to be a problem with `remark-mdx-frontmatter` 1.1.0 which is breaking the tests. My local machine had 1.0.1 but CI kept installing 1.1.0. This is also why #124 failed on merge.

Comment (Arcath) [reply package.json]: 1.1.1 fixes this issue, removed the pinning.

**Comment by Arcath** [comment]
This is looking pretty good now.

There is now a single exported function that does both files and source strings. Making it easy to add other entry methods in the future. 

Outputing the assets is not easier for end useres without having to mess around with the esbuild settings. Internally this makes it easier for us to work out if the user is wiritng to the disk. Feels a bit nicer to the end user as well I think as its our API thats handing the changes for them.

I know your working on a few things with Remix @kentcdodds  but I wonder if we should push this breaking change before the big day? I'm guessing any Remix docs that talk mdx are probabbly going to point this way.

**Review by kentcdodds — APPROVED** [review]
Yup, I'm a fan. Let's do it 👍

**Comment by AnweshGangula** [comment]
@Arcath / @kentcdodds  I'm receiving an error - `'source' or 'file' must be defined`. I've created an issue regarding the same here #149. Please let me know

**Pull request merged**